Flaky build failure on Windows (shared library configuration)

The windows shared build is failing on occasion with:

> LINK : fatal error LNK1104: cannot open file 'D:\a\shaka-packager\shaka-packager\build\packager\Release\libpackager.lib' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_static.vcxproj]

It then passes when re-run.  I think this indicates a race condition somewhere in the parallel build, but I'm not sure of the nature of it.

Here is a failed build: https://github.com/shaka-project/shaka-packager/actions/runs/7906044243/job/21580055244?pr=1340
Followed by a successful rerun: https://github.com/shaka-project/shaka-packager/actions/runs/7906044243/job/21585322587?pr=1340

## Failed build

Here are log snippets from the failed build:

**Building libpackager_shared**

```
Wed, 14 Feb 2024 19:24:15 GMT
     Creating library D:/a/shaka-packager/shaka-packager/build/packager/Release/libpackager.lib and object D:/a/shaka-packager/shaka-packager/build/packager/Release/libpackager.exp
Wed, 14 Feb 2024 19:24:15 GMT
LINK : warning LNK4217: symbol '??0Status@shaka@@QEAA@XZ (public: __cdecl shaka::Status::Status(void))' defined in 'job_manager.obj' is imported by 'media_replicator.lib(replicator.obj)' in function '"private: virtual class shaka::Status __cdecl shaka::media::Replicator::Process(class std::unique_ptr<struct shaka::media::StreamData,struct std::default_delete<struct shaka::media::StreamData> >)" (?Process@Replicator@media@shaka@@EEAA?AVStatus@3@V?$unique_ptr@UStreamData@media@shaka@@U?$default_delete@UStreamData@media@shaka@@@std@@@std@@@Z)' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_shared.vcxproj]
Wed, 14 Feb 2024 19:24:15 GMT
LINK : warning LNK4217: symbol '?Update@Status@shaka@@QEAAXV12@@Z (public: void __cdecl shaka::Status::Update(class shaka::Status))' defined in 'status.lib(status.obj)' is imported by 'media_replicator.lib(replicator.obj)' in function '"private: virtual class shaka::Status __cdecl shaka::media::Replicator::Process(class std::unique_ptr<struct shaka::media::StreamData,struct std::default_delete<struct shaka::media::StreamData> >)" (?Process@Replicator@media@shaka@@EEAA?AVStatus@3@V?$unique_ptr@UStreamData@media@shaka@@U?$default_delete@UStreamData@media@shaka@@@std@@@std@@@Z)' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_shared.vcxproj]
Wed, 14 Feb 2024 19:24:15 GMT
LINK : warning LNK4217: symbol '??1Status@shaka@@QEAA@XZ (public: __cdecl shaka::Status::~Status(void))' defined in 'job_manager.obj' is imported by 'media_replicator.lib(replicator.obj)' in function '"int `private: virtual class media::Status __cdecl shaka::media::Replicator::Process(class std::unique_ptr<struct shaka::media::StreamData,struct std::default_delete<struct shaka::media::StreamData> >)'::`1'::dtor$1" (?dtor$1@?0??Process@Replicator@media@shaka@@EEAA?AVStatus@3@V?$unique_ptr@UStreamData@media@shaka@@U?$default_delete@UStreamData@media@shaka@@@std@@@std@@@Z@4HA)' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_shared.vcxproj]
Wed, 14 Feb 2024 19:24:15 GMT
LINK : warning LNK4217: symbol '??0Status@shaka@@QEAA@AEBV01@@Z (public: __cdecl shaka::Status::Status(class shaka::Status const &))' defined in 'job_manager.obj' is imported by 'media_replicator.lib(replicator.obj)' in function '"private: virtual class shaka::Status __cdecl shaka::media::Replicator::InitializeInternal(void)" (?InitializeInternal@Replicator@media@shaka@@EEAA?AVStatus@3@XZ)' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_shared.vcxproj]
Wed, 14 Feb 2024 19:24:15 GMT
LINK : warning LNK4217: symbol '?OK@Status@shaka@@2V12@B (public: static class shaka::Status const shaka::Status::OK)' defined in 'status.lib(status.obj)' is imported by 'media_replicator.lib(replicator.obj)' in function '"private: virtual class shaka::Status __cdecl shaka::media::Replicator::InitializeInternal(void)" (?InitializeInternal@Replicator@media@shaka@@EEAA?AVStatus@3@XZ)' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_shared.vcxproj]
Wed, 14 Feb 2024 19:24:15 GMT
  libpackager_shared.vcxproj -> D:\a\shaka-packager\shaka-packager\build\packager\Release\libpackager.dll
```

**Building libpackager_static (fails)**

```
Wed, 14 Feb 2024 19:24:56 GMT
LINK : fatal error LNK1104: cannot open file 'D:\a\shaka-packager\shaka-packager\build\packager\Release\libpackager.lib' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_static.vcxproj]
```

## Successful build (rerun)

Here are log snippets from the successful build (rerun):

**Building libpackager_shared**

```
Wed, 14 Feb 2024 21:50:07 GMT
     Creating library D:/a/shaka-packager/shaka-packager/build/packager/Release/libpackager.lib and object D:/a/shaka-packager/shaka-packager/build/packager/Release/libpackager.exp
Wed, 14 Feb 2024 21:50:07 GMT
LINK : warning LNK4217: symbol '??0Status@shaka@@QEAA@XZ (public: __cdecl shaka::Status::Status(void))' defined in 'job_manager.obj' is imported by 'media_replicator.lib(replicator.obj)' in function '"private: virtual class shaka::Status __cdecl shaka::media::Replicator::Process(class std::unique_ptr<struct shaka::media::StreamData,struct std::default_delete<struct shaka::media::StreamData> >)" (?Process@Replicator@media@shaka@@EEAA?AVStatus@3@V?$unique_ptr@UStreamData@media@shaka@@U?$default_delete@UStreamData@media@shaka@@@std@@@std@@@Z)' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_shared.vcxproj]
Wed, 14 Feb 2024 21:50:07 GMT
LINK : warning LNK4217: symbol '?Update@Status@shaka@@QEAAXV12@@Z (public: void __cdecl shaka::Status::Update(class shaka::Status))' defined in 'status.lib(status.obj)' is imported by 'media_replicator.lib(replicator.obj)' in function '"private: virtual class shaka::Status __cdecl shaka::media::Replicator::Process(class std::unique_ptr<struct shaka::media::StreamData,struct std::default_delete<struct shaka::media::StreamData> >)" (?Process@Replicator@media@shaka@@EEAA?AVStatus@3@V?$unique_ptr@UStreamData@media@shaka@@U?$default_delete@UStreamData@media@shaka@@@std@@@std@@@Z)' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_shared.vcxproj]
Wed, 14 Feb 2024 21:50:07 GMT
LINK : warning LNK4217: symbol '??1Status@shaka@@QEAA@XZ (public: __cdecl shaka::Status::~Status(void))' defined in 'job_manager.obj' is imported by 'media_replicator.lib(replicator.obj)' in function '"int `private: virtual class media::Status __cdecl shaka::media::Replicator::Process(class std::unique_ptr<struct shaka::media::StreamData,struct std::default_delete<struct shaka::media::StreamData> >)'::`1'::dtor$1" (?dtor$1@?0??Process@Replicator@media@shaka@@EEAA?AVStatus@3@V?$unique_ptr@UStreamData@media@shaka@@U?$default_delete@UStreamData@media@shaka@@@std@@@std@@@Z@4HA)' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_shared.vcxproj]
Wed, 14 Feb 2024 21:50:07 GMT
LINK : warning LNK4217: symbol '??0Status@shaka@@QEAA@AEBV01@@Z (public: __cdecl shaka::Status::Status(class shaka::Status const &))' defined in 'job_manager.obj' is imported by 'media_replicator.lib(replicator.obj)' in function '"private: virtual class shaka::Status __cdecl shaka::media::Replicator::InitializeInternal(void)" (?InitializeInternal@Replicator@media@shaka@@EEAA?AVStatus@3@XZ)' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_shared.vcxproj]
Wed, 14 Feb 2024 21:50:07 GMT
LINK : warning LNK4217: symbol '?OK@Status@shaka@@2V12@B (public: static class shaka::Status const shaka::Status::OK)' defined in 'status.lib(status.obj)' is imported by 'media_replicator.lib(replicator.obj)' in function '"private: virtual class shaka::Status __cdecl shaka::media::Replicator::InitializeInternal(void)" (?InitializeInternal@Replicator@media@shaka@@EEAA?AVStatus@3@XZ)' [D:\a\shaka-packager\shaka-packager\build\packager\libpackager_shared.vcxproj]
Wed, 14 Feb 2024 21:50:08 GMT
  libpackager_shared.vcxproj -> D:\a\shaka-packager\shaka-packager\build\packager\Release\libpackager.dll
```

**Building libpackager_static (works)**

```
Wed, 14 Feb 2024 21:50:45 GMT
  libpackager_static.vcxproj -> D:\a\shaka-packager\shaka-packager\build\packager\Release\libpackager.lib
```

## Possible solutions

We could find the root of this MSVC failure (timing?) and resolve it.  @cosmin also suggests that we shouldn't be building two library configurations at once, so maybe we could restructure the targets so that we only ever build libpackager_static or libpackager_shared, but not both.

--- a/packager/CMakeLists.txt
+++ b/packager/CMakeLists.txt
@@ -112,36 +112,23 @@ set(libpackager_deps
 )
 
 # A static library target is always built.
-add_library(libpackager_static STATIC ${libpackager_sources})
-target_link_libraries(libpackager_static ${libpackager_deps})
 
-# And always installed as libpackager.a / libpackager.lib:
-if(NOT MSVC)
-  set_property(TARGET libpackager_static PROPERTY OUTPUT_NAME packager)
+if(BUILD_SHARED_LIBS)
+  add_library(libpackager SHARED ${libpackager_sources})
+  target_link_libraries(libpackager ${libpackager_deps})
+  target_compile_definitions(libpackager PUBLIC SHAKA_IMPLEMENTATION)
 else()
-  set_property(TARGET libpackager_static PROPERTY OUTPUT_NAME libpackager)
+  add_library(libpackager STATIC ${libpackager_sources})
+  target_link_libraries(libpackager ${libpackager_deps})
 endif()
 
-# A shared library target is conditional (default OFF):
-if(BUILD_SHARED_LIBS)
-  add_library(libpackager_shared SHARED ${libpackager_sources})
-  target_link_libraries(libpackager_shared ${libpackager_deps})
-  target_compile_definitions(libpackager_shared PUBLIC SHAKA_IMPLEMENTATION)
-
-  # And always installed as libpackager.so / libpackager.dll:
-  if(NOT MSVC)
-    set_property(TARGET libpackager_shared PROPERTY OUTPUT_NAME packager)
-  else()
-    set_property(TARGET libpackager_shared PROPERTY OUTPUT_NAME libpackager)
-  endif()
-
-  # If we're building a shared library, this is what the "libpackager" target
-  # aliases to.
-  add_library(libpackager ALIAS libpackager_shared)
+# The library is always installed as
+# libpackager.so / libpackager.dll / libpackager.a / libpackager.lib:
+if(NOT MSVC)
+  # The "lib" prefix is implied outside of MSVC.
+  set_property(TARGET libpackager PROPERTY OUTPUT_NAME packager)
 else()
-  # If we're not building a shared library, the "libpackager" target aliases to
-  # the static library.
-  add_library(libpackager ALIAS libpackager_static)
+  set_property(TARGET libpackager PROPERTY OUTPUT_NAME libpackager)
 endif()
 
 add_executable(packager
@@ -256,7 +243,7 @@ install(DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/pssh-box-protos
 # The static library isn't usable as a standalone because it doesn't include
 # its static dependencies (zlib, absl, etc).
 if(BUILD_SHARED_LIBS)
-  install(TARGETS libpackager_shared)
+  install(TARGETS libpackager)
   install(DIRECTORY ../include/packager
     DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
   install(FILES ${CMAKE_CURRENT_BINARY_DIR}/packager.pc

